runner/sidecar/source/db/db.go

```diff
@@ -0,0 +1,217 @@
+						continue
+					}
+					offset = fmt.Sprintf("%v", d[x.OffsetColumn])
+				}
+			}
+		}
+	}()
+
+	// update offset in db
+	go func() {
+		ticker := time.NewTicker(x.CommitInterval.Duration)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				if offset != "" {
+					if _, err := updateOffset(ctx, db, x.Table, consumerGroup, offset); err != nil {
+						logger.Error(err, "failed to update offset", "source", sourceName)
+						continue
+					}
+				}
+			}
+		}
+	}()
+
+	return dbSource{
+		db: db,
+	}, nil
+}
+
+func (d dbSource) Close() error {
+	return d.db.Close()
+}
+
+func getOffsetFromDB(ctx context.Context, db *sql.DB, tableName, consumerGroup string) (string, error) {
+	var offset string
+	if err := db.QueryRowContext(ctx,
+		"select offset from argo_dataflow_offsets where table_name=? and consumer_group=?",
+		tableName, consumerGroup).Scan(&offset); err != nil {
+		return "", err
+	}
+	return offset, nil
+}
+
+func updateOffset(ctx context.Context, db *sql.DB, tableName, consumerGroup, offset string) (int64, error) {
+	stmt, err := db.Prepare("update argo_dataflow_offsets set offset=? where table_name=? and consumer_group=?")
+	if err != nil {
+		return 0, fmt.Errorf("failed to prepare offset update statement: %w", err)
+	}
+	defer func() { _ = stmt.Close() }()
+	if result, err := stmt.ExecContext(ctx, offset, tableName, consumerGroup); err != nil {
+		return 0, fmt.Errorf("failed to exec offset update statement: %w", err)
+	} else {
+		return result.RowsAffected()
+	}
+}
+
+func insertOffset(ctx context.Context, db *sql.DB, tableName, consumerGroup, offset string) (int64, error) {
+	stmt, err := db.Prepare("insert into argo_dataflow_offsets (table_name, consumer_group, offset) values (?, ?, ?)")
+	if err != nil {
+		return 0, fmt.Errorf("failed to prepare offset insert statement: %w", err)
+	}
+	defer func() { _ = stmt.Close() }()
+	if result, err := stmt.ExecContext(ctx, tableName, consumerGroup, offset); err != nil {
+		return 0, fmt.Errorf("failed to exec offset insert statement: %w", err)
+	} else {
+		return result.RowsAffected()
+	}
+}
+
+func queryData(ctx context.Context, db *sql.DB, tableName, offsetColumn, offset string) (result []rowData, err error) {
+	sql := "select * from " + tableName + " order by " + offsetColumn
+	params := []interface{}{}
+	if offset != "" {
+		sql = "select * from " + tableName + " where " + offsetColumn + " > ? order by " + offsetColumn
+		params = append(params, offset)
+	}
+	rows, err := db.QueryContext(ctx, sql, params...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query data from table %s: %w", tableName, err)
+	}
+	defer func() { _ = rows.Close() }()
+
+	columns, err := rows.Columns()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get table columns: %w", err)
+	}
+	count := len(columns)
+	result = make([]rowData, 0)
+	values := make([]interface{}, count)
+	valuePtrs := make([]interface{}, count)
+	for rows.Next() {
```

Comment:
The result could be very large. Can we stream the data somehow?

Reply:
yes.